File: Algorithm/Java/src/main/java/com/yugutou/charpter5_queue_map/level1/LinkQueue.java
package com.yugutou.charpter5_queue_map.level1;

public class LinkQueue {
    private Node front;
    private Node rear;
    private int size;

    public LinkQueue() {
        this.front = new Node(0);
        this.rear = new Node(0);
    }

    /**
     * 入队
     */
    public void push(int value) {
        Node newNode = new Node(value);
        Node temp = front;
        while (temp.next != null) {
            temp = temp.next;
        }
        temp.next = newNode;
        rear = newNode;
        size++;
    }

    /**
     * 出队
     */
    public int pull() {
        if (front.next == null) {
            System.out.println("队列已空");
        }
        Node firstNode = front.next;
        front.next = firstNode.next;
        size--;
        return firstNode.data;
    }

    /**
     * 遍历队列
     */
    public void traverse() {
        Node temp = front.next;
        while (temp != null) {
            System.out.print(temp.data + "\t");
            temp = temp.next;
        }
    }

    static class Node {
        public int data;
        public Node next;

        public Node(int data) {
            this.data = data;
        }
    }

    //测试main方法
    public static void main(String[] args) {
        LinkQueue linkQueue = new LinkQueue();
        linkQueue.push(1);
        linkQueue.push(2);
        linkQueue.push(3);
        System.out.println("第一个出队的元素为:" + linkQueue.pull());
        System.out.println("队列中的元素为:");
        linkQueue.traverse();
    }
}

File: Algorithm/c_or_cpp/charpter5_queue_hash/level1/link_queue.cpp
#include <stdlib.h>
#include <stdio.h>

#include <stdio.h>

typedef struct Node {
    int data;
    struct Node* next;
} Node;

typedef struct LinkQueue {
    Node* front;
    Node* rear;
    int size;
} LinkQueue;

LinkQueue* createLinkQueue() {
    LinkQueue* queue = (LinkQueue*)malloc(sizeof(LinkQueue));
    queue->front = NULL;
    queue->rear = NULL;
    queue->size = 0;
    return queue;
}

void push(LinkQueue* queue, int value) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = value;
    newNode->next = NULL;
    
    if (queue->rear == NULL) {
        queue->front = newNode;
        queue->rear = newNode;
    } else {
        queue->rear->next = newNode;
        queue->rear = newNode;
    }
    
    queue->size++;
}

int pull(LinkQueue* queue) {
    if (queue->front == NULL) {
        printf("队列已空\n");
        return 0;
    }
    
    Node* firstNode = queue->front;
    int value = firstNode->data;
    queue->front = firstNode->next;
    free(firstNode);
    queue->size--;
    
    if (queue->front == NULL) {
        queue->rear = NULL;
    }
    
    return value;
}

void traverse(LinkQueue* queue) {
    Node* temp = queue->front;
    while (temp != NULL) {
        printf("%d\t", temp->data);
        temp = temp->next;
    }
}

int main() {
    LinkQueue* queue = createLinkQueue();
    
    push(queue, 1);
    push(queue, 2);
    push(queue, 3);
    
    traverse(queue);  // 输出: 1 2 3
    
    int value = pull(queue);
    printf("\n出队的元素: %d\n", value); // 输出: 1
    
    traverse(queue);  // 输出: 2 3
    
    return 0;
}